mlir/include/mlir/IR/PatternMatch.h

```diff
@@ -511,6 +512,50 @@ class RewriterBase : public OpBuilder {
     OpBuilder::Listener *listener;
   };
 
+  /// A listener that emits all notifyMatchFailure calls
+  /// with the given DiagnosticSeverity (default is Error).
+  /// This can be used so that any notifyMatchFailure calls will be emitted
+  /// as a message with the given severity (Remark will be used for Note)
+  struct MatchFailureEmittingListener : public RewriterBase::Listener {
```

Comment:
Do we actually need this? Match failures are printed during a greedy pattern rewrite / dialect conversion with `mlir-opt -debug`. Are there any other use cases for this?

Reply:
Some flows, especially downstreams built entirely around patterns, increasingly use "notifyMatchFailure" as a generic silenceable error-reporting mechanism. It's a creative misuse of the infra, but they've got nothing else.

Reply:
>  But if all that applyPartialConversion returns is LogicalResult, then how else do you describe to the user that a lowering step failed?

`applyPartialConversion` has an 4th argument to collect the operation that couldn't be legalized, which allows you to diagnose after the fact if you'd like. You could print the list of the operations that couldn't be legalized if you'd like.

Reply:
If you emit an error/warning/remark every time a pattern fails to match, you will be bombarded with output in many cases (even though the lowering was successful). Just try running `-canonicalize` on a larger piece of IR and see how many pattern fail to match. (Most of them don't report match failures due to laziness though.)